PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -136,14 +157,91 @@ UpdateLoadedImage (
 
     // Save other binary blobs
     Index = 3;
-    while ((Index < MAX_MULTIBOOT_MODULE_NUMBER) && (Index < NumFiles)) {
+    while ((Index < MAX_EXTRA_FILE_NUMBER) && (Index < NumFiles)) {
+      // Update ACPI tables if we encounter an ACPI blob
+      if (File[Index].Name == SIGNATURE_32('A', 'C', 'P', 'I')) {
+          DEBUG ((DEBUG_INFO, "Loading boot image ACPI tables...\n"));
+          PlatformService = (PLATFORM_SERVICE *) GetServiceBySignature (PLATFORM_SERVICE_SIGNATURE);
+          if ((PlatformService != NULL) && (PlatformService->AcpiTableUpdate != NULL)) {
+            Status = PlatformService->AcpiTableUpdate (File[Index].Addr, File[Index].Size);
+            DEBUG ((DEBUG_INFO, "Updating ACPI table with boot image %d - %r\n", Index, Status));
+          }
+          FreeImageData (&File[Index]);
+          Index++;
+          continue;
+        }
       CopyMem (&LinuxImage->ExtraBlob[Index - 3], &File[Index], sizeof (IMAGE_DATA));
+
+      //
+      // Update the blob's address in the kernel command line so that the OS knows where it resides
+      // We also copy the blob into a reserved memory address so that the OS does not overwrite it
+      //
+
+      // Get the blob name
+      CopyMem(BlobName, &File[Index].Name, 4);
+      BlobName[4] = '\0';
+
+      // Copy the extra blob into reserved memory
+      BlobReservedBuf = AllocateReservedPages(EFI_SIZE_TO_PAGES(File[Index].Size));
+      if (BlobReservedBuf == NULL) {
+        return EFI_OUT_OF_RESOURCES;
+      }
+      CopyMem(BlobReservedBuf, File[Index].Addr, File[Index].Size);
+      DEBUG ((DEBUG_INFO, "Copied %a to reserved memory @ 0x%016X\n", BlobName, BlobReservedBuf));
+
+      // Generate the search string: SBL.XXXX=0x0000000000000000
+      AsciiSPrint(BlobSearchStr, 28, "SBL.%a=0x0000000000000000", BlobName);
+      DEBUG ((DEBUG_INFO, "Searching for \"%a\" blob placeholder string  in cmdline: %a... ", BlobName, BlobSearchStr));
```

Comment:
Does the kernel/OS understand this format (SBL.XXXX=0x0000000000000000) to parse and extract the address?

Reply:
This info is consumed by device drivers. Typically, drivers look for binary blobs in the file system. This provides an alternate way for drivers to get some config blobs (instead of waiting for the file system to be mounted). So would require specific drivers to be aware of this.